267: Enable deletion via UI for projects/originals/sets/translations

toolstack opened the pull request:
This PR allows an admin to delete any of the various GlotPress items:
- Projects - will delete all content in the project (originals, translation sets, translations, glossaries, sub projects, etc.).
- Glossaries - will delete the glossary and all entries.
- Translation set - will delete the set and all translations.

Test cases are included.

New "delete" pages are included.

Resolves #267.

toolstack committed:
Re-enable the project delete routes.

toolstack committed:
Remove the JavaScript warning on the project delete link.

toolstack committed:
Add the project delete page.

toolstack committed:
Add the delete_post() method.

Update delete_get() method to display project delete page.

toolstack committed:
Add the delete() method to project things.

This will delete child projects, translation sets and originals associated with the project as well.

toolstack committed:
Add a translation set delete method.

This will delete all translations associated with the set as well.

toolstack committed:
Update breadcrumbs to include complete path from root.

toolstack committed:
Add the translation set delete page.

toolstack committed:
Add translation set delete routes.

toolstack committed:
Add the delete_post() method.

Add delete_get() method to display project delete page.

toolstack committed:
Added gp_link_set_delete*() helper functions.

toolstack committed:
Added the "delete" link to the translation set page.

toolstack committed:
Fixed typo in variable name.

toolstack committed:
Replaced incorrect $project with correct $set.

toolstack committed:
Fix typo in variable name and use !empty() instead of truthy.

toolstack committed:
Add method scope.

toolstack committed:
Add delete() method to glossary thing.

toolstack committed:
Delete glossaries as part of deleting a translation set.

toolstack committed:
Add glossary delete template file.

toolstack committed:
Add glossary delete routes.

toolstack committed:
Add the delete_post() method.

Add delete_get() method to display glossary delete page.

toolstack committed:
Added gp_link_glossary_delete*() helper functions.

toolstack committed:
Updated incorrect variable name.

toolstack committed:
Add glossary delete link.

toolstack committed:
Fix typo in variable name.

toolstack committed:
Declare scope and remove old comments.

toolstack committed:
Added test cases.

toolstack committed:
Added test cases for translation set route for delete_post().

toolstack committed:
Replaced hidden inputs with route variable.

toolstack committed:
Added delete_many() method and doc blocks.

toolstack committed:
Converted individual delete() calls to delete_many() calls.

toolstack committed:
Declare scope of method.

ocean90 commented on gp-templates/project-delete.php:
```diff
@@ -0,0 +1,16 @@
+gp_breadcrumb_project( $project );
+gp_tmpl_header();
+?>
+<h2><?php echo wptexturize( sprintf( __( 'Delete project "%s"', 'glotpress' ), esc_html( $project->name ) ) ); ?></h2>
```
To avoid `wptexturize()` you add the correct quotes to the string: `'Delete project &#8220;%s&#8221;'`.

toolstack replied:
Copied straight from the edit template:

https://github.com/GlotPress/GlotPress-WP/blob/develop/gp-templates/project-edit.php#L6

Is it better to remain consistent and then create a PR to solve all of them or update this one instance?

ocean90 replied:
Just update this instance.

toolstack committed:
Replace use of wptexturize with embedded html entitles.

ocean90 commented:
Some feedback:
- Use a new cap `delete` instead of existing write/approve caps
- Split after_delete into a new PR
- Doc blocks
- No `javascript:history.back();`

toolstack committed:
Remove use of javascript back().

Also update the glossary breadcrumbs to use gp_url_join() instead of a string addition.

toolstack committed:
Changed 'write' permission to 'delete' for deleting.

toolstack committed:
Add doc blocks.

toolstack committed:
Added doc blocks.

toolstack committed:
Add return type to doc block.

toolstack committed:
Updated doc block.

toolstack committed:
Update parameter to be the path instead of the id.

toolstack committed:
Check the right variable and condition when deleting a project.

ocean90 commented on gp-includes/things/project.php:
```diff
@@ -342,5 +342,22 @@ public function duplicate_project_contents_from( $source_project ){
+		
+		GP::$translation_set->delete_many( array( 'project_id' => $this->id ) );
+
+		GP::$original->delete_many( array( 'project_id' => $this->id ) );
```
Shouldn't each step check whether the previous delete process was successful?

toolstack replied:
I'd say no, simply because we're not going to undo the previous steps if something fails.

There could also be valid instances of a step "failing" (like maybe the project doesn't have any glossaries or translation sets) but we would still want to continue on.

ocean90 committed:
Impove docs and fix coding style issues.

Also, fix syntax for `customAutoEscapedFunctions` in `phpcs.ruleset.xml`.
